File: Iceland/100/iceland.vinir.cpp
#include<bits/stdc++.h>
using namespace std;

int par[100050];
int sz[100050];

int n, q;

int uf_find(int a){
	if(par[a] == a) return a;
	else {
		par[a] = uf_find(par[a]);
		return par[a];
	}
}

bool uf_union(int a, int b){
	int pa = uf_find(a);
	int pb = uf_find(b);
	if(pa != pb){
		par[pb] = uf_find(a);
		sz[pa] += sz[pb];
		return true;
	} else {
		return false;
	}
}

int uf_size(int a){
	return sz[uf_find(a)];
}

int main(){
	cin >> n >> q;

	for(int i = 1; i <= n; i++) par[i] = i, sz[i] = 1;

	for(int i = 0; i < q; i++){
		int op;
		cin >> op;
		if(op == 1){
			int a, b;
			cin >> a >> b;
			uf_union(a, b);
		} else {
			int a;
			cin >> a;
			cout << uf_size(a) - 1 << endl;
		}
	}
}

File: Iceland/100/iceland.vinir.py
par = [0]
sz = [1]

def uf_find(a):
    if par[a] == a: return a
    else:
        par[a] = uf_find(par[a])
        return par[a]

def uf_union(a, b):
    pa = uf_find(a)
    pb = uf_find(b)
    if pa != pb:
        par[pb] = uf_find(a)
        sz[pa] += sz[pb]
        return True
    else:
        return False

def uf_size(a):
    return sz[uf_find(a)]

def main():
    n, q = map(int, input().split())

    for i in range(1, n+1):
        par.append(i)
        sz.append(1)

    for i in range(q):
        op = list(map(int, input().split()))

        if op[0] == 1:
            a, b = op[1], op[2]
            uf_union(a, b)
        else:
            a = op[1]
            print(uf_size(a)-1)

main()
